### DISCUSSION TITLE
Custom, styled loglevel prefixes without markup=True

### DISCUSSION THREAD

**Discussion by dionhaefner** [Textualize/rich#2798] [Q&A] [answered]
Hey,

I am using rich to style log messages which works beautifully for the most part, but I'm having issues with getting styled custom loglevel prefixes without enabling `RichHandler(..., markup=True)` (which I don't want to do because I have log messages coming from untrusted sources).

Example code:

```python
import logging
import logging.handlers
from rich.logging import RichHandler

LEVEL_PREFIX = {
    "DEBUG": " \\[d] ",
    "INFO": " [bold]\\[+][/] ",
    "WARNING": " [yellow]\\[!][/] ",
    "ERROR": " [red]\\[-][/] ",
    "CRITICAL": " [red bold]\\[x][/] ",
}

package_logger = logging.getLogger(__name__)
package_logger.setLevel("DEBUG")  # send everything to handlers

# stream handler with custom formatter
fmt = "{levelprefix!s}{message!s}"
ch = RichHandler(level="DEBUG", rich_tracebacks=True, show_time=False, show_level=False, show_path=False)

class PrefixFormatter(logging.Formatter):
    def format(self, record, *args):
        # register custom field
        record.levelprefix = LEVEL_PREFIX.get(record.levelname, "")
        return super().format(record, *args)

ch_fmt = PrefixFormatter(fmt, style="{")

ch.setFormatter(ch_fmt)
package_logger.handlers = [ch]

package_logger.warning("hi there!")
```

Which prints

```
 [yellow]\[!][/] hi there!
```

as expected.

Is there any way to apply styling only to the prefixes?

**Comment by jmw182** [comment]
I would like to do something similar. I'd like to add the logger name to the format string, and apply custom styling such as making it bold or applying a color to separate it from the log message string. 

Something like:
```python
formatter = logging.Formatter("{name} {message}", style="{")
handler = RichHandler()
handler.setFormatter(formatter)
``` 

with custom styling applied to `"{name}"`. This can work with `markup=True`, but what I'd really like is a way to apply rich markup to the Formatter string without applying it to the contents of each log message.

**Reply by dionhaefner** [reply, marked as answer]
I solved this for my use case by calling `rich.markup.escape` inside a custom formatter:

```python
from rich.markup import escape

class PrefixFormatter(logging.Formatter):
    def format(self, record: Any, *args: Any) -> Any:
        record.levelprefix = LEVEL_PREFIX.get(record.levelname, "")
        record.msg = escape(record.msg)
        return super().format(record, *args)
```

That way you can use `markup=True` but ignore control codes in messages.

**Reply by jmw182** [reply]
Thanks @dionhaefner, this works great for me!
